students/scott_bromley/lesson09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,246 @@
+            raise ValueError(f"{self.donor_name} has no donations")
+        return round(mean(self.donations), 2)
+
+    def add_donation(self, donation):
+        """
+        adds donation to Donor donations
+        :param self:
+        :param donation: donation amount
+        :return: None
+        """
+        if donation <= 0:
+            raise ValueError("invalid donation amount")
+        else:
+            try:
+                self.donations.append(float(donation))
+            except TypeError as type_err:
+                raise type_err("invalid donation type")
+
+    def thank_you(self):
+        """
+        create formatted letter string for donor
+        :return: formatted letter
+        """
+        if not self._donations:
+            raise ValueError("donor has no donations")
+        else:
+            return f"\n\n{datetime.today().strftime('%Y-%d-%m')},\n\n\n" \
+                   f"Dear {self.donor_name},\n\n\n" \
+                   f"On behalf of all of us we thank you for your generous\n" \
+                   f"donation of ${self.donations[-1]:03.2f}.\n" \
+                   f"To date, you have donated: ${self.total_donations:03.2f}.\n" \
+                   "We look forward to seeing you at our annual meeting.\n\n" \
+                   "Sincerely,\n\n" \
+                   "ADL\n"
+
+    def report_row(self):
+        """
+        format report row for Donor object
+        :return: formatted report row Donor
+        """
+        if not self._donations:
+            raise ValueError("donor has no donations")
+        else:
+            return ('{:<25}{:^5}${:>14,.2f}{:^5}{:>10}{:^5}${:>14,.2f}'
+                    ).format(self.donor_name, ' ',
+                             self.total_donations, ' ',
+                             self.num_donations, ' ',
+                             self.avg_donation
+                             )
+
+    def __str__(self):
+        """
+        prints Donor object
+        :return: String representation of Donor object
+        """
+        return f"Donor: {self._donor_name}, Donations: {self._donations}"
+
+    def __repr__(self):
+        """
+        :return: string of self instantiation
+        """
+        return f"Donor({self._donor_name!r}, {self._donations!r})"
+
+    @staticmethod
+    def sort_key(self):
+        return self.last_name, self.first_name, self.donations
+
+    @staticmethod
+    def sort_by_donations(self):
+        return self.donations
+
+
+class DonorCollection:
+    """
+    DonorCollection class is a collection of Donor objects
+    """
+    def __init__(self, *donors):
+        """
+        DonorCollection
+        :param donors: iterable of Donor object(s)
+        """
+        self._donors = defaultdict(Donor)
+        for donor in donors:
+            try:
+                self._donors[donor.donor_name].append(donor.donations)
+            except TypeError as donor_err:
+                raise donor_err("object is not of type Donor")
```

Comment:
I don't see code where you're passing arguments to `DonorCollection` can you add that to git so I can see? Otherwise I can't tell what error you would be getting since you're not using it anywhere in the files here.

Your code above does have an issue though. You are creating a defaultdict and say that "if I don't set the value for my dictionary I want you to use Donor object as a default".. then you are appending to a dictionary value like you expect the value to be a list. What your intention was, is to have a name (string) for a key and Donor object as a value (not a list).

so what you want instead is:
```
for donor in donors:
    self._donors[donor.name] = donor
```